### DISCUSSION TITLE
How to fix laravel api cors error

### DISCUSSION THREAD

**Discussion by rahman-cmd** [community/community#112267] [Programming Help]
### Select Topic Area

Question

### Body

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Display JSON Data</title>
</head>
<body>
    <div id="json-data"></div>

    <script>
        // Fetch JSON data from the API endpoint
        fetch('https://site.co/api/v1/customer/config')
            .then(response => response.json())
            .then(data => {
                // Render the JSON data on the HTML page
                const jsonDataElement = document.getElementById('json-data');
                jsonDataElement.textContent = JSON.stringify(data, null, 2);
            })
            .catch(error => {
                console.error('Error fetching data:', error);
            });
    </script>
</body>
</html>

Access to fetch at 'https://site.co/api/v1/customer/config' from origin 'null' has been blocked by CORS policy: The 'Access-Control-Allow-Origin' header contains multiple values '*, *', but only one is allowed. Have the server send the header with a valid value, or, if an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

**Comment by Dione-b** [comment]
> <title>Display JSON Data</title>
> ```
> <script>
>     // Fetch JSON data from the API endpoint
>     fetch('https://site.co/api/v1/customer/config')
>         .then(response => response.json())
>         .then(data => {
>             // Render the JSON data on the HTML page
>             const jsonDataElement = document.getElementById('json-data');
>             jsonDataElement.textContent = JSON.stringify(data, null, 2);
>         })
>         .catch(error => {
>             console.error('Error fetching data:', error);
>         });
> </script>
> ```
> 
> Access to fetch at 'https://site.co/api/v1/customer/config' from origin 'null' has been blocked by CORS policy: The 'Access-Control-Allow-Origin' header contains multiple values '*, *', but only one is allowed. Have the server send the header with a valid value, or, if an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

The error message you're encountering is related to Cross-Origin Resource Sharing (CORS), a security feature implemented in web browsers to restrict web pages from making requests to a different domain than the one that served the web page. This is a common issue when trying to fetch data from an API that doesn't allow requests from your origin.

There are a few ways to address this issue:

1. Modify Server Configuration
The most straightforward solution is to modify the server configuration to include the correct Access-Control-Allow-Origin header. This header tells the browser which origins are allowed to access the resource. If you have control over the server, you can set this header to allow requests from your origin. For example, if your site is served from https://example.com, you would set the header to Access-Control-Allow-Origin: https://example.com.

2. Use a Proxy Server
If you don't have control over the server, you can use a proxy server that adds the necessary CORS headers to the response. This involves making your request to the proxy server, which then makes the request to the API on your behalf and adds the CORS headers to the response before sending it back to your site. There are many proxy services available, or you can set up your own.

3. Use a CORS Plugin or Extension
For development purposes only, you can use a CORS plugin or extension for your browser that allows you to bypass CORS restrictions. This is not recommended for production use, as it poses security risks.

4. Use no-cors Mode
As the error message suggests, you can set the request's mode to no-cors to fetch the resource with CORS disabled. However, this results in an "opaque" response, which means you won't be able to read the data or view the status of the response. This is generally not useful for most applications.

Example Using no-cors Mode
Here's how you could modify your fetch request to use no-cors mode. Note that this will not allow you to read the response data:

```<title>Display JSON Data</title>
<script>
    // Fetch JSON data from the API endpoint with no-cors mode
    fetch('https://site.co/api/v1/customer/config', {
        mode: 'no-cors' // Set the mode to 'no-cors'
    })
    .then(response => {
        // Since the response is opaque, you cannot read the data
        console.log('Fetch completed, but data cannot be read due to CORS restrictions.');
    })
    .catch(error => {
        console.error('Error fetching data:', error);
    });
</script>
```
The best solution depends on your specific situation. If you're developing a production application, modifying the server configuration or using a proxy server are the most appropriate solutions. For development purposes, using a CORS plugin or extension can be a quick workaround, but it's not suitable for production.

**Comment by amankr1469** [comment]
`    fetch('https://site.co/api/v1/customer/config', {
        mode: 'no-cors'
    })`
    
    This disables CORS checks (but the approach is not recommended for most cases due to security concerns).

**Comment by AbdulVakeel** [comment]
Install a Laravel CORS package like fruitcake/laravel-cors using Composer.

Publish the CORS package configuration using php artisan vendor:publish --tag="cors".

Configure CORS settings in the config/cors.php file according to your requirements.

Register the CORS middleware in app/Http/Kernel.php by adding 'cors' => \Fruitcake\Cors\HandleCors::class to the middleware array.

Apply the CORS middleware to your API routes or groups of routes where you want CORS to be enabled using the middleware method.

Clear the configuration cache with php artisan config:cache.

Test your API endpoints to ensure CORS error #112267 is resolved.
#112267

**Comment by ahmadrauf51** [comment]
### How to fix this error in our Laravel API:

To begin with, we have to install a package to handle CORS headers. A popular package is **fruitcake/laravel-cors**. We can install it with Composer.

Then, we have to register the Cors middleware in our **app/Http/Kernel.php** file.

This is done by adding it to the **$middleware** array:

```
protected $middleware = [
// ...
\Fruitcake\Cors\HandleCors::class,
];
```

Now, it is time to publish the configuration file for the laravel-cors package with the Artisan command:

`php artisan vendor:publish –tag=cors`

This command will create a **config/cors.php** configuration file.

Then, open the **config/cors.php** file and configure it as per our needs. We can mention which origins, methods, headers, and other settings are allowed for our API.

```
For example:
'paths' => ['api/*'],
'allowed_methods' => ['*'],
'allowed_origins' => ['*'],
'allowed_headers' => ['*'],
'exposed_headers' => [],
'max_age' => 0,
'supports_credentials' => false,
```

We can adjust these settings to match our needs.

At this point, we have to test our API requests from different origins to make sure that the CORS policy is correctly configured and that we can make requests without encountering CORS errors.

Finally, we can make changes to our CORS configuration for better performance. After that, we can clear the cache using Artisan:

`php artisan config:clear`

And then re-cache the configuration:

`php artisan config:cache`

Furthermore, if our API is meant to be accessible from multiple origins, we can use proper authentication and authorization mechanisms to secure our API endpoints.

**Comment by wgarciar3** [comment]
The error you're actually hitting is different from a normal "CORS not configured" problem — look closely: `'Access-Control-Allow-Origin' header contains multiple values '*, *'`. That means the header is present **twice** in the response, not missing.

This happens when CORS is being added in two places at once — most commonly:
- Laravel's own CORS handling (`fruitcake/laravel-cors` or the built-in `HandleCors` middleware / `config/cors.php`), AND
- your webserver in front of it (Nginx `add_header 'Access-Control-Allow-Origin' '*';` in the server/location block, or an `.htaccess` rule if you're on Apache).

Browsers reject a response with a duplicated CORS header even if both values are identical — only one is allowed.

Fix: pick exactly one layer to own CORS, not both.
- If you keep it in Laravel: remove any `add_header Access-Control-Allow-Origin` line from your Nginx config (check the `api`/`server` block and any `location /api { }` block — sometimes it's added in more than one place there too), then `php artisan config:clear`.
- If you'd rather handle it at the webserver level: remove Laravel's CORS middleware/package entirely and configure it only in Nginx.

Quick way to confirm before touching config: `curl -i https://site.co/api/v1/customer/config` and look at how many `Access-Control-Allow-Origin` lines come back in the raw response headers.

**Comment by nileshvishwakarma2004** [comment]
For Python 3.13, the "pkg_resources" deprecation issue is related to the current package version and its dependencies.

A practical workaround is to use Python 3.12 in a virtual environment:

py -3.12 -m venv .venv
.venv\Scripts\activate
python -m pip install -U pip setuptools
python -m pip install openai-whisper

If you specifically need Python 3.13, installing the latest Whisper code directly from GitHub can also be used as a temporary workaround:

python -m pip install git+https://github.com/openai/whisper.git

This avoids relying on the older PyPI release while compatibility is being updated.

**Comment by Zohaibcode740** [comment]
The `*, *` in the error is the key here. Your API is returning the `Access-Control-Allow-Origin` header twice.

Check both Laravel and your web server (Nginx/Apache) for CORS configuration. Keep CORS enabled in only one place and remove the duplicate header from the other.

You can confirm it with:

```bash
curl -I https://site.co/api/v1/customer/config
```

If you see `Access-Control-Allow-Origin` more than once, that confirms the issue. After fixing the duplicate configuration, clear Laravel's config cache:

```bash
php artisan config:clear
php artisan cache:clear
```

I would **not** use `mode: "no-cors"` here because you need to read the JSON response, and a no-cors response will be opaque.

**Comment by Hilal-11** [comment]
CORS is blocked by the browser, not Laravel itself. In modern Laravel (9+), it's built-in — no package needed.

1. Publish the config (if not present):

bash
php artisan config:publish cors

2. Edit config/cors.php:

php
return [
    'paths' => ['api/*', 'sanctum/csrf-cookie'],
    'allowed_methods' => ['*'],
    'allowed_origins' => ['http://localhost:3000'], // your frontend URL, not '*'
    'allowed_origins_patterns' => [],
    'allowed_headers' => ['*'],
    'exposed_headers' => [],
    'max_age' => 0,
    'supports_credentials' => true, // only if using cookies/Sanctum auth
];

3. Make sure the middleware is active — in Laravel 11+, HandleCors is already in the global middleware stack by default, so you usually don't need to add it manually.

4. Clear config cache after any change:

bash
php artisan config:clear

Common mistakes:

Origin must match exactly — http://localhost:3000 ≠ http://127.0.0.1:3000, and no trailing slash.
Don't set allowed_origins => ['*'] together with supports_credentials => true — browsers reject that combo.
If routes aren't under api/*, add your actual path prefix to the paths array.

**Comment by ducthais** [comment]
I'm here to read everyone's great answers!
